proposals/undersampling_project_proposal.md

```diff
@@ -0,0 +1,47 @@
+#### SIG TFX-Addons
+
+# Project Proposal
+
+------
+
+**Your name:** Daniel Kim
+
+**Your email:** danielk@twitter.com
+
+**Your company/organization:** Twitter
+
+**Project name:** Undersampling Component
+
+## Project Description
+
+This project will be a custom function-based component that inputs an artifact in `tfRecord` format of `tf.Example`s and randomly undersamples it, reducing the data to the lowest-frequency class. It will primarily use an underlying Apache Beam pipeline that will be wrapped inside the TensorFlow component.
+
+## Project Category
+
+Component
+
+## Project Use-Case(s)
+
+As this project represents a very general operation used widely in machine learning data processing, we anticipate that it will have wide-ranging use cases, the most evident being in cases where dependent variable classes have wildly different relative frequencies and undersampling is needed to help effectively train a classifier. The potential impact will likely be large due to this, and our organization will likely utilize this project in the future.
+
+## Project Solutions
+
+We considered multiple possible solutions and implementations for this project before deciding on an Apache Beam-based pipeline, including standard Python code and the utilization of a BigQuery query in order to perform the random undersampling task. Using a pure Python-based algorithm with `multiprocessing` will likely be inefficient for the purposes of a parallelizable computation such as this one, and utilizing solutions such as Dask would introduce unnecessary dependencies into our project. 
+
+BigQuery is also a very good option, and a great fallback in case Apache Beam turns out to be infeasible for this project, but Apache Beam has better python integration through custom `DoFn`s that may help us with our implementation of other algorithms later on. In this case, we would load the data in and out of a BigQuery table and perform our operations within this table. The component would then either utilize a schema generated from `SchemaGen` or inflect one on its own, potententially adding an unneeded dependency into the component or performing unnecessary inflection.
+
+## Project Implementation
+
+At a high level, the plan is to use Apache Beam to injest a `tfRecord` of `tf.Examples`, shuffle them, convert them into a key-value `PCollection` with keys as class values and values as data points, and then use `Sample.FixedSizePerKey()` in order to perform the actual undersampling. The algorithm will be written as an Apache Beam pipeline, which will be wrapped into a TensorFlow custom function component to use with TFX pipelines. The component would be written as inputting a `tfRecord` artifact and exporting a similar `tf.Record` artifact, making its placement in a pipeline nearly ubiquitous. If necessary, the component may also be changed to a fully-custom component, albeit one where only the executor is edited.
```

Comment:
ingest

Reply:
What's the use case of TensorFlow custom function, vs just another Python function?

Reply:
This is actually outdated, but the use of a TFX custom component function would be to allow its use in TFX pipelines as a component. The current implementation uses a fully-custom component with a custom executor and spec, which fulfills a similar role while allowing more complexity.